sphinx-mode.el

```diff
@@ -123,49 +122,54 @@ If BUFFER is not given use the `current-buffer'."
-		      (setq dir nil)
-		      nil)
-                  (not (file-exists-p (concat dir file-name))))
-        ;; Move up to the parent dir and try again.
-	(setq prevdir dir)
-        (setq dir (expand-file-name (file-name-directory
-                                     (directory-file-name
-				      (file-name-directory dir))))))
-      (or (and dir (concat dir "")) nil))))
-
-;; TODO: replace this with sphinx-compile and sphinx-compile-and-view
-(defun sphinx-view-build-html ()
-  "Runs 'make html' in project root directory and then views compiled HTML of current source file."
+    (let
+      (
+       (buffer-name-base (file-name-base (buffer-file-name)))
+       (project-root-dir (locate-dominating-file buffer-file-name "Makefile"))
+       (target-format (read-from-minibuffer "Make (default html): " nil nil nil nil "html" nil))
+       )
+      (let
+	  (
+	   (file-rel-path (file-relative-name buffer-file-name project-root-dir))
+	   )
+        (shell-command (concat "make -C " project-root-dir " " target-format)))))
+
+(defun sphinx-compile-and-view ()
+  "Runs sphinx-compile and then views compiled version of current source file."
   (interactive)
     (let
       (
-       (buffer-name-base (file-name-base (buffer-name))) ; reads name of current bufer (presumably a .rst file), minus extension, into buffer-name-base
-       (project-root-dir (sphinx-compile-find-makefile-dir)) ; finds parent directory containing makefile and reads into project-root-dir
+       (buffer-name-base (file-name-base (buffer-file-name)))
+       (project-root-dir (locate-dominating-file buffer-file-name "Makefile"))
        )
       (let
 	  (
-	   (file-rel-path (string-remove-prefix project-root-dir (file-name-directory buffer-file-name)))
+	   (file-rel-path (file-relative-name buffer-file-name project-root-dir))
 	   )
         (shell-command (concat "make -C " project-root-dir " html"))
-	(browse-url (concat (concat (concat project-root-dir "_build/html/" file-rel-path) buffer-name-base) ".html"))
-	)
-    )
-  )
+  (interactive)
+    (let
+      (
+       (buffer-name-base (file-name-base (buffer-file-name)))
+       (project-root-dir (locate-dominating-file buffer-file-name "Makefile"))
+       (target-format (read-from-minibuffer "Make (default html): " nil nil nil nil "html" nil))
+       )
+      (let
+	  (
+	   (file-rel-path (file-relative-name buffer-file-name project-root-dir))
+	   )
+	(xdg-open (expand-file-name
+		   (concat project-root-dir "_build/" target-format "/" (last (file-name-all-completions (file-name-sans-extension file-rel-path)))))))))))
 
 (defvar sphinx-mode-map
   (let ((map (make-sparse-keymap)))
     (define-key map (kbd "M-'") 'sphinx-goto-ref)
     (define-key map (kbd "C-c TAB") 'sphinx-insert-ref)
-    (define-key map (kbd "C-c C-c C-v") 'sphinx-view-build-html)
+    (define-key map (kbd "C-c C-x C-c") 'sphinx-compile)
+    (define-key map (kbd "C-c C-x C-v") 'sphinx-compile-and-view)
```

Comment:
Could we find a way to reuse `sphinx-compile` in `sphinx-compile-and-view`.  There seems to be a lot of similar code.